[Needs Feedback] Logistics Pipes Compendium

Bernasss12 opened the pull request:
### Introduction
So the plan with the first book was to be different from the original book. for reasons like readability and expand-ability and honestly because I thought this mod deserved it.
After that comes version 2.0 of said book, in this version the idea was to not really change the aspect or feel but to make the book an actual viable replacement to a online wiki (provided it is populated with information, for now I'm making the skeleton later we fill in the flesh).
Some may have seen some of the progress over on the discord, others may already be tired of my endless questions ( :P ), but for those who are not aware of this little project and stumbled across here I'll do my best to list what kinds of features this book already has and what is already decided to be added.

### Current features (with some still being WIP):
- **3 levels of organization** 
1. Division, is a group of chapters and is defined by a title and it's group of chapters;
2. Chapter, can be a normal book chapter or a list/enumeration of pages (lists WIP), also have titles and have multiple pages;
3. Page, can be displayed as a group of pages or a individual page. Stores all the text information.
- **Restores closed page when reopening book**
- **Reads all the information from a json file and an organized and localized folder structure **
- **Rudimentary bookmark system (WIP)**

### Todo list:
- Add text customization like coloring, and formatting ();
- Image support;
- Link? I belive this would be useful for page linking, not too sure about web linking (Needs Comments);
- Search bar? Not too sure of how needed this would be but if it is wanted let me know (Needs Comments)
- NEI/JEI information page integration, also needs more feedback. (Needs Comments)

[Glo Gloo](https://app.gitkraken.com/glo/board/XQp5bM0RewAQYHZ1)

Commit:
First try at making a generic guide book.

Added simple Item, still have to attach it to the event listener and add a recipe but that's not a priority.
Added a scale-able GUI that works around the already used translation based book text.
Semi-implemented slider functionality but that's also not a priority as of now.

Commit:
Resolved multiple issues

- Auto-indented before committing.
- Reworked background drawing
- Fixed everything I was asked for :P

Commit:
Added slider functionality changed some things.

- Fully implemented the slider functionality but still has no uses as no sentence fills the whole space.
- Changed the drawing functions to a more flexible one, visually the same.
- Everything should scale better now.
- Still waiting for more feedback :P

Commit:
Many changes

- Packet implementation (sort of)
- Make working slider, check it out.
- Created entry nr 11 on .lang file, used only to debug.

Commit:
I think this is a good enough version for evaluation

- All basic functionality of a normal book (and more) are implemented, now onto the good part.

Commit:
Remade some key parts

- Routed the GUI opening through the ClientProxy.
- Reformatted the packet to not use NBT but instead simple values.
- Made it work _flawlessly_ with the offhand.

Commit:
Formatting fixes

- Changed `splitLines()` method as requested :)
- Stopped using JavaDoc everywhere
- Formatted comments as they where supposed to be

Commit:
Manually Removed GuiGuideBook from GuiHandler

I thought I had reverted the class to the beginning of the branch, I guess it didn't work so there it goes :)

Commit:
Ready for Review

- Added Texture
- Added Recipe
- Added to EventListener
- Added localized name
- Removed from GuiIDs
- Renamed Proxy method
- Fixed minor bug

Commit:
Remake drawing and added book organization system.

- Remade all the drawing with custom made functions using tessellation.
- Added a couple of classes as an idea of a book system.

Commit:
Major changes to the rendering and book structure

Bernasss12 committed:
Minor change

Commit:
Auto stash before rebase of "origin/feature/custom-guide-book-experimental"

Commit:
Lots of additions

 - Reworked tiles
 - New expandable file reading structure
 - New Menu with tiles and Title boxes
 - New PageState system (not renamed yet but functioning)
 - New tile Title overlay
 - New Buttons (home, next/prev page)
and maybe some more little things I can't recall right now :)

Commit:
More additions, manly bookmarks

Added bookmarks (WIP)

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+	public int max, min;
+
+	public void setProgress(float progress){
+		this.y = (int)((this.max-this.min)*progress);
```
Is that supposed to be a "scale from min to max in a range of 0-1" type method? Then you need `min + (max - min) * progress`
You might also, you know, actually set the `progress` variable here, or the name is misleading.

Bernasss12 replied:
I think I ended up not using this method at all, but the point was to give the float progress to the button and have it situate in the right `y` level, I guess I should fix it and actually use it

bziemons replied:
also: give this calculation some space

Bernasss12 replied:
And I have have the names switched for some reason I may change it for simplicity's sake but right now max is the top limit and min is the bottom limit :/

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+		this.zLevel = z;
+		this.max = yTop;
+		this.min = yBot - heightIn;
+		this.y = (int) (max + (min - max) * progress);
```
Prevent duplicate code, call setProgress here

bziemons replied:
Also: why is this calculation so much different from the one above?

Bernasss12 replied:
Yes nice call

Bernasss12 replied:
Because one of them was wrong but I've already fixed it :S

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+
+	public void reset(){
+		this.y = max;
+		this.progress = 0.0F;
```
call setProgress

bziemons replied:
or replace the usages of this function with `setProgress(0.0F);`

Bernasss12 replied:
RIght

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+
+	@Override
+	public void playPressSound(SoundHandler soundHandlerIn) {
+		return;
```
Why?

Bernasss12 replied:
So the slider doesn't play a sound. Unless it is wanted to play a sound when pressed?

bziemons replied:
I think he means the useless return. you can just replace the method body with `{}` after the head though.

dblsaiko replied:
Nope, I did mean why override that to do nothing in the first place. Yeah, probably better if that doesn't play any sound.

Bernasss12 replied:
But sure I can remove the useless return too

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+	@Override
+	public boolean mousePressed(Minecraft mc, int mouseX, int mouseY) {
+		if(super.mousePressed(mc, mouseX, mouseY)){
+			dragging = true;
```
No need to have this, there's a field that gets set when dragging. Check super.mousePressed
(And remove the `dragging` field)

Bernasss12 replied:
`mousePressed()` only checks if the button is `visible && enabled && hovered`
`mouseDragged()` doesn't care if the mouse is over the button or not at any time

And from what I found in the Minecraft code the creative inventory slider also uses a boolean value to start and stop following the mouse, so I don't really know of any other way to do this

dblsaiko replied:
Ahh, the field that I meant was in GuiScreen, `selectedButton`. Yeah, the slider won't be able to access it. My bad.

Bernasss12 replied:
No problem

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookTabButton.java:
```diff
@@ -0,0 +1,52 @@
+			color = this.isActive?0xFFFFFF:color;
+			GuiGuideBook.drawStretchingSquare(x, drawY - drawHeight, x + width, drawY, 15, 40, 64, 40 + width, 64 + drawHeight, true, color);
+			GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+			this.mouseDragged(mc, mouseX, mouseY);
```
Why does drawButton call mouseDragged?

Bernasss12 replied:
Just because it was called in super, not sure if it is actually needed

Bernasss12 replied:
But I guess it doesn't so I removed it

dblsaiko commented on common/logisticspipes/gui/guidebook/book/MenuItem.java:
```diff
@@ -0,0 +1,127 @@
+	//// Atlas
+	// Btn
+	private final int atlas$btn$bgU0 = 64, atlas$btn$bgV0 = 32, atlas$btn$bgU1 = 96, atlas$btn$bgV1 = 64;
+	private final int atlas$btn$u0 = 0, atlas$btn$v0 = 64, atlas$btn$u1 = 2, atlas$btn$v1 = 66, atlas$btn$u2 = 14, atlas$btn$v2 = 78, atlas$btn$u3 = 16, atlas$btn$v3 = 80;
```
Don't use `$` in variable names in most cases, it's ugly and mostly used only by generated stuff

Bernasss12 replied:
Sure

dblsaiko commented on common/logisticspipes/gui/guidebook/book/MenuItem.java:
```diff
@@ -0,0 +1,127 @@
+	private GuideBookContents.Chapter chapter;
+
+	//// Drawing variables
+	public boolean visiible, hovering, enabled;
```
`visiible` → `visible`

Bernasss12 replied:
Yeah actually spotted that one while resolving the previous but thanks hahaha

dblsaiko commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -31,7 +21,7 @@
 	@Getter
 	@Setter
-	private int page;
+	private int page, chapter, division;
```
This will completely break when removing/modifying book pages. Use full page paths (as in, path to the page file), not indices

Bernasss12 replied:
I will do that in the Content getter but how would it be better to use that inside the code?

Bernasss12 replied:
Added to Glo Board

dblsaiko commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -46,11 +36,16 @@ public void processPacket(EntityPlayer player) {
+		if(book.hasTagCompound()){
+			nbt = book.getTagCompound();
+		}else{
+			nbt = new NBTTagCompound();
```
You're not setting the NBT on the item anymore if it doesn't exist

Bernasss12 replied:
So what should I check? The `ItemStack `is being grabbed only when the packet runs so in case the book doesn't exist or for whatever reason in not in the hand slot anymore it won't get to this line of the code. 
The only thing this `if` is checking is if the book already had NBT or not since if it did it should only edit it and if it didn't it can be just set.

dblsaiko replied:
That's not what I'm saying. You need to call book.setTagCompound if you're giving the item a completely new tag.

Bernasss12 replied:
Oh yes done

dblsaiko commented on common/logisticspipes/utils/GuideBookContents.kt:
```diff
@@ -8,56 +9,84 @@ import net.minecraft.util.ResourceLocation
+    }
+
+    fun getChapter(dindex: Int, index: Int): Chapter?{
+        return getDivision(dindex).getChapter(index)
```
Why are these needed? Just call getDivision().getChapter(), it's not really more code and is more readable

Bernasss12 replied:
If `getDivision().getChapter()` is more readable I guess they aren't needed, just thought having a function specifically for that purpose would be more readable, I'll remove then

dblsaiko replied:
In this case you have 2 parameters, which you won't know what they do without looking at the parameter names or the function body, and when not using this function you'd have only 1 parameter but 2 calls, but since the function names are stuff like `getDivision` it's clear what the parameter is

Bernasss12 replied:
I changed this but I'll show you why I had that in the first place:
`currentPage = new SavedTab(gbc.getDivision(nbtTagCompound.getInteger("division")).getChapter(nbtTagCompound.getInteger("chapter")).getPage(nbtTagCompound.getInteger("page")), page, nbtTagCompound.getFloat("sliderProgress"));`

dblsaiko replied:
Yeah, that's a very long line. In that case, I'd just insert some line breaks to make it look better.
```java
currentPage = new SavedTab(
  gbc.getDivision(nbtTagCompound.getInteger("division"))
    .getChapter(nbtTagCompound.getInteger("chapter"))
    .getPage(nbtTagCompound.getInteger("page")),
  page, nbtTagCompound.getFloat("sliderProgress")
);
```
Something like that.

Bernasss12 replied:
Yes I better do that :)

dblsaiko commented on common/logisticspipes/utils/GuideBookContents.kt:
```diff
@@ -8,56 +9,84 @@ import net.minecraft.util.ResourceLocation
+    }
+
+    fun getPage(dindex: Int, cindex: Int, index: Int): Page?{
+        return getDivision(dindex).getChapter(cindex)?.getPage(index)
```
Why are these needed? Just call getDivision().getChapter()?.getPage(), it's not really more code and is more readable

Bernasss12 replied:
If `getDivision().getChapter()?.getPage()` is more readable I guess they aren't needed, just thought having a function specifically for that purpose would be more readable, I'll remove then

dblsaiko commented on common/logisticspipes/utils/GuideBookContents.kt:
```diff
@@ -8,56 +9,84 @@ import net.minecraft.util.ResourceLocation
+        val nPages = json["pages"].asInt
+
+        fun getPage(index: Int): Page? {
+            var par = parentTitle.replace("[^A-z]".toRegex(), "_").toLowerCase()
```
Don't get the file path from the title, or the page will need to be renamed whenever the title changes, which is not ideal. Also this prevents consistent file names across localizations.

Bernasss12 replied:
Yes that is already in the set to be done, unfortunately I've been having problems with stuff not working so those have taken priority

Bernasss12 replied:
Added to Glo board

dblsaiko commented on common/logisticspipes/utils/GuideBookContents.kt:
```diff
@@ -8,56 +9,84 @@ import net.minecraft.util.ResourceLocation
-  }
 
-  class Page(val index: Int, val text: String)
+    class Division constructor(lang: String, json: JsonObject, val dindex: Int) {
```
Don't need `constructor` if you don't need to modify constructor visibility.

dblsaiko commented on resources/assets/logisticspipes/book/en_us/book.json:
```diff
@@ -1,4 +1,72 @@
+        {
+          "title": "Quickstart Guide",
+          "item": "minecraft:apple",
+          "pages": "10"
```
Instead of "title" and "pages" in the main file, put them in the directory with all the pages, so a chapter is self-contained.
And of course, add a path to the chapter here instead.

bziemons replied:
why is pages a String?

Bernasss12 replied:
So having a .json file for the book and one for each chapter? Sounds good but I guess I need to have everything that is not working fixed first :/

Bernasss12 replied:
Added that to my Glo board

Bernasss12 replied:
The page is a string for no reason to be honest when calling into the Object it gets turned into an int

dblsaiko replied:
Why is this (and others) marked as resolved if it, well, isn't? When you fix it with a commit, it autocollapses, until then, keep comments open so that you can actually see what is resolved and what isn't, please.

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+			GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+			this.hovered = mouseX >= this.x && mouseY >= this.y && mouseX < this.x + this.width && mouseY < this.y + this.height;
+			int i = this.hovered&&!dragging||!this.enabled?1:0;
+			int j = this.dragging||!this.enabled?1:0;
```
Replace these with booleans with readable names, and then use the `? :` in the line below to make it more readable

Bernasss12 replied:
Yes right I was hacking away at the time sry about that

Bernasss12 replied:
Not sure if what I did was exactly what you meant but I'm sure it helps readability

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookTexturedButton.java:
```diff
@@ -0,0 +1,104 @@
+		this.sizeX = 0;
+		this.sizeY = 0;
+		this.hasDisabledState = false;
+		this.type = EnumButtonType.TAB;
```
Instead of setting every field here again, call `this(buttonId, x, y, widthIn, heightIn, 0, 0, …)` instead

Bernasss12 replied:
Done

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookTexturedButton.java:
```diff
@@ -0,0 +1,104 @@
+			if (sizeX > 0 && sizeY > 0)
+				drawTexturedButtonForegroundLayer(mc, mouseX, mouseY, GUI_BOOK_TEXTURE, u1, v1, sizeX, sizeY);
+		}else if(this.type == EnumButtonType.NORMAL){
+			GlStateManager.pushMatrix();
```
You're not modifying the transformation matrix, so why push/pop? (In case you are doing that in drawTexturedButtonForegroundLayer, call push/pop in there and not outside)

Bernasss12 replied:
This was before I knew what the transformation matrix stored

dblsaiko commented on common/logisticspipes/gui/guidebook/GuiGuideBookTexturedButton.java:
```diff
@@ -0,0 +1,104 @@
+			this.hovered = mouseX >= this.x && mouseY >= this.y && mouseX < this.x + this.width && mouseY < this.y + this.height;
+			int i = (!this.hasDisabledState && this.hovered) || (this.hasDisabledState && this.enabled && hovered)? 1 : 0;
+			int j = (this.hasDisabledState && !this.enabled) ? 2 : 0;
+			int offSetX = (this.width - sizeX) / 2;
```
"offset" is a single word, so `offsetX` and `offsetY` would be better names

dblsaiko reviewed (changes requested):
I've said this before, but format your code before committing. It greatly increases readability and you should always do it anyways even if this is not going to land in the main repo just yet.

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+import logisticspipes.gui.guidebook.book.MenuItem;
+import logisticspipes.utils.GuideBookContents;
+
+public class GuiGuideBook extends GuiScreen {
```
Would you be okay with this being licensed under MIT as well? If yes, please move it to the `network.rs485...` package then and have the default copyright generated (you should replace RS485 with your name of course).

Bernasss12 replied:
Could you explain whats the point? I don't mind if it is licensed under MIT but I don't understand the package change

Bernasss12 replied:
Added to the Glo Board

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+public class GuiGuideBook extends GuiScreen {
+
+	/*
+	 * TODO - Back button or bookmarks, whichever is more wanted
```
Please no TODOs in code

Bernasss12 replied:
Yeah I've already set up a project so I'll remove that, it's the Glo I put in the first comment

bziemons replied:
Sadly I seem to need to login to see the Glo. If you want us to see it, you could post/update a screenshot of that

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+	private static final ResourceLocation GUI_BOOK_TEXTURE = new ResourceLocation(LPConstants.LP_MOD_ID, "textures/gui/guide_book.png");
+
+	private GuideBookContents gbc;
+	//////// Buttons
```
Please don't fancy style your comments, but keep the usual coding style

Bernasss12 replied:
Sure

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+	 */
+
+	private boolean loadedNBT = false;
+	private final int z$tooltip = 20;      // Tooltip z
```
> [...] Additionally, the dollar sign character, by convention, is never used at all. You may find some situations where auto-generated names will contain the dollar sign, but your variable names should always avoid using it. [...]

quoted from https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+
+	//////// Drawing variables
+	//// Gui
+	private int gui$x0, gui$y0, gui$x1, gui$y1, gui$x2, gui$y2, gui$x3, gui$y3, gui$width, gui$height, gui$innerWidth, gui$innerHeight, gui$bgX0, gui$bgY0, gui$bgX1, gui$bgY1;
```
see above.

Use camelCase like `guiX0`, `guiHeight` instead

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+	 */
+	protected void getDataFromNBT() {
+		if (loadedNBT) return;
+		ItemStack bookItemStack = mc.player.getHeldItem(hand);
```
From looking at this method, one shouldn't expect that the book is in the player's hand. Please pass the ItemStack as parameter and have the `mc.player.getHeldItem(hand)` in an appropriate spot

Bernasss12 replied:
Why wouldn't we expect the book to be in the players hand? How would he activate it otherwise?

bziemons replied:
well, at least check if the ItemStack's item is actually the book. Can you not at all switch items while the GUI is open? What happens if there was a slight lag?

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+
+	@Override
+	protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException {
+		if (state == menu)
```
leaving away curly brackets is okay with me for one-liners with good visibility, but this certainly isn't the case here.

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+					if (savedTabs.addTab(new SavedTab(setPage, state))) {
+						tab.isSet = true;
+						if (tab.isSet) tab.isActive = true;
+					}else{
```
Give the poor *else* some space, duh

Bernasss12 replied:
I need to `Ctrl+Alt+L` more often I know I now x)

dblsaiko replied:
I pretty much press that every time I stop typing. There's also a "format on commit" option in IntelliJ's commit dialog, but that will format every file you've touched, even the parts of it that you didn't touch, so it's not a good idea in this project. I do need to make a "format everything" commit at some point though, so that that won't be a problem anymore.

Bernasss12 replied:
I have to develop the habit

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+					setPage.progress = currentPage.progress;
+					if (savedTabs.addTab(new SavedTab(setPage, state))) {
+						tab.isSet = true;
+						if (tab.isSet) tab.isActive = true;
```
eh?

Bernasss12 replied:
I pretty sure I changed this last night, didn't push because other problems emerged T.T

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+					}else{
+						return;
+					}
+				} else if(tab.isSet && !tab.isActive){
```
code style, spaces

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+	protected void updateTitle() {
+		String title = "";
+		title += gbc.getTitle();
+		if (state == menu) title += ": Menu";
```
if you have an else with brackets, use brackets on the if, please

Bernasss12 replied:
Surely :)

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+		this.nextPage.enabled = currentPage.getIndex() < currentPage.getPageCount() - 1;
+		this.nextPage.visible = state == page;
+		this.home.visible = state != menu;
+		for(GuiGuideBookTabButton tab: tabList){
```
code style, spaces

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+		updateButtonVisibility();
+	}
+
+	/* *********************************************** draw functions with a twist *********************************************** */
```
fancy comments again, eh?

Bernasss12 replied:
I'm a fancy guy I guess

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+
+	/* *********************************************** draw functions with a twist *********************************************** */
+
+	/*
```
If you have a doc on a method, please use JavaDoc notation then: `/** ... */`

Bernasss12 replied:
Hummm so can I actually use JavaDoc notation on these ones? I just avoided it altogether

dblsaiko replied:
If you have documentation for a specific method/class/field, then yes, use JavaDoc comments. They won't work for comments that are supposed to target a section of methods, like the "draw functions" comment here, since they will only apply to the method directly following the comment.

Bernasss12 replied:
Makes sense, and tbh I only used to use JavaDoc everywhere because it was green text and normal comments were very hard to see, I have changed IDEA since but yeah :/

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+	 * If the texture size is smaller than the draw size the texture will be repeated until it fills all the area to be drawn.
+	 * The Y increases from the top to the bottom.
+	 */
+
```
remove unnecessary new-lines

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+		}
+	}
+
+	public interface State {
```
Don't know if *State* is the correct title for this interface. Looks more like a simple `IDrawable`?

Bernasss12 replied:
I guess you are right, I'll leave all the naming for you guys to be honest

Bernasss12 replied:
What would the implementations be called?

bziemons replied:
The rest was fine, I only meant this interface. Will have a look at the names in my next review though

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+		return a.page.getPage() == b.page.getPage() && a.page.getChapter() == b.page.getChapter() && a.page.getDivision() == b.page.getDivision() && a.state == b.state;
+	}
+
+	public class MenuItemsDivision {
```
The file has 600 lines until this point. Please consider using multiple files for additional classes. Also consider the visibility of those classes. Do they all have to be *public*?

Bernasss12 replied:
Added to the Glo Board

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBook.java:
```diff
@@ -0,0 +1,779 @@
+
+	public class SavedTabs {
+
+		public ArrayList<SavedTab> tabs;
```
Consider making this private and re-create the methods you need. Like `public int size() { return tabs.size() }`

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+	private boolean dragging;
+	public int max, min;
+
+	public void setProgress(float progress){
```
Why is *setProgress* not setting `this.progress`?

bziemons replied:
also code style, spacing

Bernasss12 replied:
Because it was only half made and unused :( But I fixed it and gave it some purpose in life

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+			mc.getTextureManager().bindTexture(GUI_BOOK_TEXTURE);
+			GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+			this.hovered = mouseX >= this.x && mouseY >= this.y && mouseX < this.x + this.width && mouseY < this.y + this.height;
+			int i = this.hovered&&!dragging||!this.enabled?1:0;
```
code style, spacing

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookSlider.java:
```diff
@@ -0,0 +1,87 @@
+
+	@Override
+	public boolean mousePressed(Minecraft mc, int mouseX, int mouseY) {
+		if(super.mousePressed(mc, mouseX, mouseY)){
```
code style, spacing

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookTabButton.java:
```diff
@@ -0,0 +1,52 @@
+
+public class GuiGuideBookTabButton extends GuiButton {
+
+	private static final ResourceLocation GUI_BOOK_TEXTURE = new ResourceLocation(LPConstants.LP_MOD_ID, "textures/gui/guide_book.png");
```
I have seen `GUI_BOOK_TEXTURE` several times now. I think initializing it once and using the same constant over several classes would be better?

Bernasss12 replied:
Also added to the Glo Board

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookTabButton.java:
```diff
@@ -0,0 +1,52 @@
+
+	private static final ResourceLocation GUI_BOOK_TEXTURE = new ResourceLocation(LPConstants.LP_MOD_ID, "textures/gui/guide_book.png");
+
+	public boolean isSet, isActive;
```
existing code rarely to never uses definition of multiple properties in one line. Please use existing code style.

bziemons replied:
with an additional note: in the rendering code above, I am fine with it. Although I am not yet sure, if you need that many variables.

Bernasss12 replied:
It eases the thinking but I could try to use less in the future

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookTabButton.java:
```diff
@@ -0,0 +1,52 @@
+
+	@Override
+	public void drawButton(Minecraft mc, int mouseX, int mouseY, float partialTicks) {
+		if (this.visible)
```
rather:
if (!this.visible) return;

Bernasss12 replied:
That would work too indeed

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookTabButton.java:
```diff
@@ -0,0 +1,52 @@
+			mc.getTextureManager().bindTexture(GUI_BOOK_TEXTURE);
+			int drawHeight = height;
+			int drawY = y;
+			if(!hovered && !isActive){
```
code style, spacing

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookTexturedButton.java:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	public void drawTexturedButtonForegroundLayer(Minecraft mc, int mouseX, int mouseY, ResourceLocation resource, int u0, int v0, int sizeX, int sizeY) {
+		if (visible) {
```
again: if (!this.visible) return;

bziemons commented on common/logisticspipes/gui/guidebook/GuiGuideBookTexturedButton.java:
```diff
@@ -0,0 +1,104 @@
+			int offSetX = (this.width - sizeX) / 2;
+			int offSetY = (this.height - sizeY) / 2;
+			GuiGuideBook.drawStretchingSquare(this.x + offSetX, this.y + offSetY, this.x + offSetX + sizeX, this.y + offSetY + sizeY, (int) zLevel + 1, u0, v0 + (sizeY) * i + (sizeY) * j, u0 + sizeX, v0 + sizeY + (sizeY) * i + (sizeY) * j);
+			this.mouseDragged(mc, mouseX, mouseY);
```
and again calling `mouseDragged` ?

Bernasss12 replied:
No need, removed

bziemons commented on common/logisticspipes/gui/guidebook/book/MenuItem.java:
```diff
@@ -0,0 +1,127 @@
+	private static final ResourceLocation GUI_BOOK_TEXTURE = new ResourceLocation(LPConstants.LP_MOD_ID, "textures/gui/guide_book.png");
+
+	// Getting constants
+	private final int z$titleButtons = 15; // Title and Buttons Z
```
variable names, see above

bziemons commented on resources/assets/logisticspipes/book/en_us/guides/quickstart_guide/page0:
```diff
@@ -0,0 +1,2 @@
+"It's been a long time. How have you been?"
+First off, thanks for playing Logistics Pipes. A lot of changes have been made since the last release of LP in Minecraft 1.7.10. We'll try to explain some of these in this book. Read the next pages, if you'd like to know them.
```
Please compact the quickstart guide to two or three pages

Bernasss12 replied:
Hahah sure

bziemons commented on common/logisticspipes/gui/guidebook/book/MenuItem.java:
```diff
@@ -0,0 +1,127 @@
+	public void drawMenuItemFrame(Minecraft mc, int mouseX, int mouseY, int x, int y, int sizeX, int sizeY) {
+		mc.renderEngine.bindTexture(GUI_BOOK_TEXTURE);
+		{
+			btn$bgX0 = x + 1;
```
why do you need a class property for this? it is only set to `x + 1` here and then used in the function call below. Please check this with all class variables for the next review round.

Bernasss12 replied:
Its just to be easier for me to spot rendering problems, and I've had plenty of those, once all the difficult part is dealt with I'll go around removing useless variables

bziemons reviewed (changes requested):
I have mostly covered visual changes now. I will re-review when those are done.

bziemons commented:
Great PR! Good explanation of general structure and what it solves and what it isn't in the opening post. Also I may forward a compliment for it by a friend and fellow software engineer.

PS:

> others may already be tired of my endless questions ( :P ),

Don't feel like that, please :) Besides that I don't know anyone who is annoyed or anything like that.

Bernasss12 committed:
Lots of requested changes

Bernasss12 commented:
I don't think it is that great, I believe this is the minimum someone could do when trying to add to a project, explain what it does, why it does it and how to make it do what it does. But I do understand that sometimes this just doesn't happen

Bernasss12 committed:
Finished adding bookmarks

Bernasss12 commented:
### Update
In the latest commit I completed the (volatile) bookmark system as well as fixed some graphical and functional bugs.

Next up I'll be adding a way to save the bookmarks in the item's NBT, it may take some time since this is looking very complicated to me and may take a while to get right.

Bernasss12 committed:
NBT Bookmarks (WIP)

Not working atm

Bernasss12 commented:
### Added in the last commits:
Color changing in bookmark tabs, Middle-click to change to next color, Shift+Middle-click to change to previous color. May be changed in the future.

It would be very nice to have the current features stress tested, and the newest changes (as said in the commit message) reviewed :)

Commit:
Oops last two commits where actually broken...

Same change log

Commit:
Separated the IDrawables interface and classes from main Gui Class

Commit:
Formatting fixes and removed multiple repeated texture loads.

Commit:
Implemented markdown tags for the most basic text formats

Bernasss12 commented:
### Added since the last commit:

**Bookmarks** aka **Tabs** are fully functional, you can add up to 10 of them, change their color over the 16 colors seen throughout all Minecraft's colored blocks (wool, glass, etc.), these tabs will be saved in the Item's `NBT `tag (so if you happen to need more than 10 bookmarks just craft up another book) and you can also remove these tabs. Here are the controls, and pardon me for repeating myself in some:

- Add tabs by clicking the plus sign above the `GUI`.
- Remove tabs by `Ctrl+Shift+Left-Clicking` on top of them
- Cycle onto the next color by `Middle-Clicking` a tab.
- Cycle back to the previous color by `Shift+Middle-Clicking` a tab.

**Text Formatting** using **Markdown**, sort of:

- Write in bold text by placing `*` around the area.
- Write in italics by placing `_` around the area.
- Write with a underline by placing `__` under the area.
- Write with underline placing `~~` around the text.

These cannot be added together, only used separately.
Also, it is not a feature but a reminder that you can use `§[0-F]` to color text, and this can be used with the previous formatting options.

Noradly committed:
Add a config option for the number of modules

in a chassis pipe. Closes #1321

bziemons commented:
Just opened the book in my dev environment and got the following crash after I added `~~you~~` as a first line in the first page. Don't know if that's related though.

[crash-2019-07-08_15.26.09-client.txt](https://github.com/RS485/LogisticsPipes/files/3368555/crash-2019-07-08_15.26.09-client.txt)

Bernasss12 commented:
Apparently it broke before reaching that phase, did you also change the book.json?

Bernasss12 commented:
My guess is that the NBT saved page is non-existent since there isn't a npe check, and there really should be one tbh

bziemons commented:
> did you also change the book.json?

No

Bernasss12 commented:
I'm not home atm and will only be able to thoroughly check later today, but I'm almost certain the problem is not the `~~you~~` because the formatting is a simple replacing of the "markdown expression" to its Minecraft counterpart

hron84 commented on common/logisticspipes/utils/string/StringUtils.java:
```diff
@@ -187,10 +189,11 @@ public static String insertThousandsSeparators(String source) {
 			boolean header = paragraph.indexOf("##") > -1;
 			if (header) paragraph = paragraph.replace("##", "");
-			if (paragraph.charAt(paragraph.length() - 1) != '\n') paragraph += '\n';
+			if (paragraph.length() != 0 && paragraph.charAt(paragraph.length() - 1) != '\n') paragraph += (TextFormatting.WHITE + "\n");
```
@Bernasss12 if the `paragraph` is null, then `paragraph.length()` will generate NPE. 

Suggested solution:

```
if (paragraph != null && 
    paragraph.length() != 0 && 
    paragraph.charAt(paragraph.length() - 1) != '\n') {
        paragraph += (TextFormatting.WHITE + "\n");
}
```

Bernasss12 replied:
I don't know how a null paragraph would come from a `split()` but I guess it's better to be safe than sorry

hron84 replied:
Honestly I didn't checked the source of the string array, but later maybe you refactor this code and feed paragraphs from different sources (like JSON parser or NBT parser or anything else) and at that point you will be thankful for this.

Bernasss12 replied:
All of this part of the code will drastically change in the near future so this was just a quick fix so people can start populating the book itself, but yes thank you for that catch

bziemons replied:
I disagree. `paragraph` isn't null here as it's String methods are used the two lines above that line. Adding the null check here adds confusion in any case. Assuming that the `paragraphs` array does not contain null values is absolutely valid and any refactoring should check for null before inserting into the array then.

The general rule should be to not use null and mark all other code with `@Nullable` from the `javax.annotation` package. Just to clarify this bit. I understand that the code will change.

Bernasss12 committed:
Fixed NPE and added color reset at the end of each paragraph.


Fixed NPE and added color reset at the end of each paragraph.

bziemons commented on common/logisticspipes/utils/string/StringUtils.java:
```diff
@@ -6,6 +6,8 @@
 import java.util.Iterator;
 import java.util.List;
 
+import javax.xml.soap.Text;
```
remove that

Bernasss12 replied:
Don't know how I added that but removed it now

Bernasss12 committed:
Removed unused imports

bziemons commented:
@Bernasss12 How does the latest commit solve the NPE, I experienced?

Bernasss12 commented:
> 
> 
> @Bernasss12 How does the latest commit solve the NPE, I experienced?

Assuming your problem was the same that **suchtie** had (as you said) the `NPE` was being thrown when the program tried to check the last `char` on an empty `String` and now I added a check to see if the `String.length()` is different than zero, meaning it has at least one `character` (in theory), as well as an actual null check before that, making the `charAt()` check safe to perform without `NPE` being a possibility, I think this makes sense 😖

bziemons commented:
```
java.lang.NullPointerException: Unexpected error
	at logisticspipes.gui.guidebook.book.SavedTab.getDivision(SavedTab.java:33)
	at logisticspipes.gui.guidebook.GuiGuideBook.updateTitle(GuiGuideBook.java:334)
	at logisticspipes.gui.guidebook.GuiGuideBook.func_73866_w_(GuiGuideBook.java:230)
	at net.minecraft.client.gui.GuiScreen.func_146280_a(GuiScreen.java:478)
	at net.minecraft.client.Minecraft.func_147108_a(Minecraft.java:1017)
	at logisticspipes.proxy.side.ClientProxy.openGuideBookGui(ClientProxy.java:258)
	at logisticspipes.items.ItemGuideBook.func_77659_a(ItemGuideBook.java:38)
	at net.minecraft.item.ItemStack.func_77957_a(ItemStack.java:217)
	at net.minecraft.client.multiplayer.PlayerControllerMP.func_187101_a(PlayerControllerMP.java:511)
```
^ that is the crashlog suchtie send me. And mine is just a few comments above.

Bernasss12 commented:
Oh that one, he did send me that crash report but then he said it was his fault, I ended up not looking at it, sorry... But probably a few null checks are in order I guess...

Bernasss12 committed:
Fix NPE when stored page does not exist anymore

Only fixed the last open page, saved tabs that no longer exist will still crash the game.

Bernasss12 commented:
Okay, please verify the changes I made on `getDataFromNBT()` and I think the problem _should_ be solved...

Bernasss12 commented:
Note that opening non-existing tabs and non-populated chapters or divisions still crashes the game.

bziemons committed:
Update to ForgeGradle 3

bziemons committed:
Generate sources jar with classifier

without overriding plain jar.

bziemons committed:
Enable processResources without mcversion

bziemons committed:
Modify CI configurations, remove setup

bziemons committed:
Configure all dependencies compileOnly

bziemons committed:
Use the manifest for build variables

Re-generate the run configurations with ForgeGradle 3.
Move the non-constants out of LPConstants.

bziemons committed:
Re-add downloading languages

-- this took way too long for a simple task..

bziemons committed:
Revert using getter for tabLabel; Fix #1391

reverts part of de30990255d8b838e6a8cd8b5dbb99a92c032679

bziemons committed:
Finalize jar by reobfJar for distribution

bziemons committed:
Throw away dev env helper and include CCL AT

The dev env now fully works \o/

bziemons committed:
Make LP build again

Remove progwml6 maven repo (timeout)
Make JEI version sticky

bziemons committed:
Update jarRepositories.xml

bziemons committed:
Restructure sourceSets

bziemons committed:
Update BUILDING.md

bziemons committed:
Re-add generated .name file

bziemons committed:
Refactor IRouterManager.getOrCreateRouter

bziemons committed:
Refactor nullability for various getRouter methods

bziemons committed:
Show enabled log levels in pre-init

Bernasss12 committed:
Change draws in StringBasedItemSink. Fix #1433 (#1437)

Changed the draw calls in GuiStringBasedItemSink.java and GuiOreDictItemSink.java because drawing the rectangles with no blend would override all the text.

Bernasss12 commented:
### Yes, this is still alive (barely).
I'm working on the file structure for the book and I have two different opinions I would like to get some feedback on if someone is interested:

## Option A
```
book/
├── de_de/
│   ├── chapter_1/
│   │   ├── page_1.md
│   │   ├── page_2.md
│   └── chapter_2/
│       └── page_1.md
├── en_us/
│   ├── chapter_1/
│   │   ├── page_1.md
│   │   ├── page_2.md
│   │   └── page_3.md
│   └── chapter_2/
│       ├── page_1.md
│       ├── page_2.md
│       └── page_3.md
└── pt_pt/
│   └── chapter_1/
│       └── page_1.md
└── book.json
```

So this option consists of one main `book.json` file that will contain the whole structure of the book, from chapter names, chapter contents as well as other information. The exact structure of the file is not important for this.
Each different "page" I stored in a markdown file for ease of formatting.
Each language has it's own folder.
This was my first idea when I restarted this project and I thought of this because this way when the book is updated it is updated to every language and even if some languages are missing some page translations, it can default to English and still have that information available for better of for worse.

## Option B
```
book/
├── de_de/
│   ├── chapter_start/
│   │   ├── first_stages.md
│   │   ├── page_2.md
│   └── chapter_end_game/
│   │   └── page_1.md
│   └──── book.json
├── en_us/
│   ├── chapter_1/
│   │   ├── page_1.md
│   │   ├── page_2.md
│   │   └── page_3.md
│   └── chapter_2/
│   │   ├── page_1.md
│   │   ├── page_2.md
│   │   └── page_3.md
│   └──── book.json
└── pt_pt/
    ├── chapter_101/
    │   └── that_page.md
    └──── book.json
```
Each language has it's own folder as well as it's own `book.json` file.
This way the translated folders don't have to follow the main book structure.

This is kind of how it works right now but I do think it has a shortcoming because it's a given that most translations will be incomplete (if they even exist) in comparison to English and using this structure it will mean that when that does happen the nonexistent files will just not be listed in the `book.json` file and therefore not accessible at all.

This is somewhat like how Wikipedia works in the sense that each page's translation can be completely different from the other languages in what comes to structure, this option is way more flexible that the previous option.

### I'd love to get other people's opinions on this, even if you think completely differently from both of these examples. Share your ideas because this is a community project and I don't feel like it's my place to just decide how to structure this based only on my opinion because I won't be the only one using it (hopefully).

bziemons committed:
Reorder, refactor ServerRouter

bziemons committed:
Remove (very) old SearchNode

bziemons commented:
> I'd love to get other people's opinions on this, even if you think completely differently from both of these examples. Share your ideas because this is a community project and I don't feel like it's my place to just decide how to structure this based only on my opinion because I won't be the only one using it (hopefully).

Just decide whichever you like more. I personally would rather like option B because I think if someone is writing a guide in their language they shouldn't have to stick to the English version. Translating that much does come with a few problems and people might just stop half way and never publish their translation. So if they can look at the English version and then publish their own text, it might be more accepted.

Bernasss12 committed:
Multiple GL11 flag resets. Fix #1441 (#1450)

* Multiple GL11 flag resets. Fix #1441

Reset color and flags with GlStateManager and GL11#glEnable/glDisable in multiple points in  both the base LogisticsBaseGuiScreen.java class as well as the GuiCraftingPipe.java class to atone for some GlStateManager leaks that have yet to be identified.
This will fix the current problem but this is not a complete fix, this is just a simple hack.

* Fixed Fuzzy tooltip alignment.

Bernasss12 committed:
Remove run configurations #1451

as they can be generated and should be customizable.

Bernasss12 commented:
After a long discussion session with @bziemons we decided on the following structure:
The whole book will be comprised solely of markdown files, with special metadata in them that will link them and identify them, and will be individually translated with mean the structure for a different language doesn't have to follow the English language even though it _can_.
This may sound relatively confusing at first but here are some more details, at the top of each .md file should be a HTML comment with YAML metadata to identify itself and it's structure. Here is an example for the main menu:

```
<!---
title: Logistics Pipes' Compendium
menu:
  Guides: 
    - /guides/*
  Item Sinks:
    - /items/itemsinks/oredict_itemsink.md
--->
```

What this would mean is that the _translated_ title for this page when displayed will be "Logistics Pipes' Compendium" and it will have some tiles that will link to the linked pages found in "menu".

Those tiles will get the icon and title information from those linked files themselves as opposed to the old main .json file that was the backbone of the whole book and could (and probably would) get very convoluted and hard to work on at some point. Then from those pages there could be other linked pages either via a list of buttons (still to be thought about) or hyperlinked text.

bziemons committed:
Use Java 1.8 output for Kotlin compiler

bziemons committed:
Refactor getRouterUnsafe -> getServerRouter

and remove useless IRouterManager interface.

Bernasss12 committed:
Fixing up the rebase.

bziemons committed:
Refactor getSpecificInterests to collector method

bziemons committed:
Remove duplicated code in ServerRouter

bziemons committed:
Refactor sneaky modules, add SneakyDirection

Rename GuiExtractor which makes it clearer it is being used by the advanced extractor as well.

bziemons committed:
Remove LogisticsSimpleFilterModule

add SimpleFilter, remove duplicate code.

bziemons committed:
Move sneaky direction read/write to interface

bziemons committed:
Make getGui methods static in SimpleFilter

bziemons committed:
Annotate param for NBT read/write non-null

bziemons committed:
Replace Nullable in SimpleFilterInventorySlot

bziemons committed:
Rename chassis slots array, reword config

and sort array of chasses slots to be ascending automatically.

bziemons committed:
Refactor packets, extract getWorldTile

Remove weird targetNotFound exception function.

bziemons committed:
Add getWorld to isServer call

bziemons committed:
Replace null-check with isEmpty

bziemons committed:
Name slot parameter and add non-null annotations

bziemons committed:
Mark chunk dirty on chassis gui close

Fixes https://github.com/RS485/LogisticsPipes/pull/1321#issuecomment-515676376

bziemons committed:
Add default implementation for getSubModule

bziemons committed:
Remove unused CrafterDefault packet

bziemons committed:
Move getSubModule to chassis pipe class

bziemons committed:
Rename getTile to getTileAs, remove duplicate code

Refactor packets and GUI providers

bziemons committed:
Pin versions, get rid of CoFH maven

bziemons committed:
Refactor LogisticsModule, remove getX/Y/Z

Add getBlockPos to get the pipe's BlockPos

bziemons committed:
Remove IntegerCoordinates and unused methods

Bernasss12 committed:
New BookContent object, added kotlinx serialization to the env.

bziemons committed:
Remove LogisticsGuiModule, add interface

Refactor all the things \o/

bziemons committed:
Create first async helpers

AsyncRouting has access to the ServerRouter and provides a update routing
table coroutine as well as direct access to the ServerRouter.

AsyncLogisticsManager currently implements the getDestination function from
the LogisticsManager with some boost on the getBestReply function as well as
an allDestinations function that returns a Kotlin Flow. Flows are based on
the Reactive Streams concept.

bziemons committed:
Add AsyncModule

all the refactoring around LogisticsModule finally makes some sense!

This just does some tick calculation and runs the first Kotlin coroutine with
async in the GlobalScope. Sadly this means there is no "structured concurrency"
which Kotlin (and me) prefers. Setup and Completion are done from the tick
method and stack traces work surprisingly well. Generics give the setup and
async result type-safety. everyNthTick is a property that can be overridden.

bziemons committed:
Cleanup and preparation for async modules

bziemons committed:
Add asynchronous extractor module (in the code)

a few tests will have to be made for this to replace the regular
extractor module, but it looks very promising. Right now it does
everything the regular extractor module does, but asynchronous!

Bernasss12 committed:
Added new BookContent object class.

Added new BookContent class that will be responsible for the guidebook parsing.
All the classes that need to interact with it were refactored and are now all working fine together.
Some of the refactored classes have been moved into the rs845 package and are now under the MIT license.
Some classes just had to have the imports updated.
Unused classes were removed.

Bernasss12 committed:
Added support for other menu formats. This should be easily built upon.

Bernasss12 committed:
Made a little environment to test the current book parsing method.

Bernasss12 committed:
Added a check for pages with no text to not break.

bziemons committed:
Only check adjacent in server thread and refactor

... IRouter, pull methods to ServerRouter.

bziemons committed:
Add resource copy script

bziemons committed:
Register world provider on chassis as well

Fixes crash with chassis pipe neighbor block change.

bziemons committed:
Add runnable in module to chunk synchronous work

bziemons committed:
Add sequencing and timeouts to async extractor

ChunkedChannel will take care of looping until its hasWork returns
false. It will take care of Channel sending and closing. Exception
handling will make sure the Channel is always closed with the caught
exception.

On the module side the sequence is limited by time usage and is chunked
into 4 entries each iteration (it should be fair to assume that
calling getStackInSlot less than 4 times will not cause measurable lag).

bziemons committed:
Fix #1465

bziemons committed:
Add fast exit when attached inv has zero size

bziemons committed:
Remove takeWhileServerNotOverloaded

.. it was a dream. Will have to implement our own time tracker for
ticks to make it work though.

Bernasss12 committed:
A bit more name and package refactoring.

Bernasss12 committed:
Added the Tokenizer and Tokens

bziemons committed:
Add toString to LogisticsModule

bziemons committed:
Move all ticks to the end phase

and dont tick twice o.O

bziemons committed:
Add timeout to async modules...

Add fast return conditions
Fix looping infintely in async extractor module

bziemons committed:
Check work before channeling sequence

... in ChunkedChannel.

Bernasss12 committed:
Created a Rectangle object to help with ui calculation

Bernasss12 committed:
Started to standardize menu widgets

bziemons committed:
Require Forgelin 1.8.4

Bernasss12 committed:
Tweaked the BookContents and page structure

The Yaml information on the pages will now be a Map with string keys and Map values.
Each entry will represent a possible button set on the page. Each entry's map is the just like it used to be, string keys and list of strings for values.
In the markdown part of the page there will be a markdown entry where you need to specify the first level key of the map and the kind button set it will be (tiles or list).
It should look something like this for example: $[map_key](type=menu)

bziemons committed:
Reset upgrade count for extraction upgrades

Fixes #1431 #1297

Bernasss12 committed:
Added the TokenMenu option

This in combination with the previous commit should allow the existence of multiple menu's within the same page, just like you would add an image or an header.

Bernasss12 committed:
Updated test files with the new Token/Menu formatting

bziemons committed:
Empty dummy build.gradle

bziemons committed:
Add Kotlin compiler options

Set language version
Allow @JvmDefault

bziemons committed:
Declare shared manifest attributes

bziemons committed:
Add AsyncQuicksortModule

bziemons committed:
Fix skipping every second slot in async quick sort

bziemons committed:
Move updateRoutingTable to tickAsync

bziemons committed:
Have allDestinations handle the routers to exclude

bziemons committed:
Rename and move AsyncLogisticsManager

because there is no async left in here

bziemons committed:
Run filter before getDestination

bziemons committed:
Replace previous LogisticsManager hasDestination

with the new LogisticsManager's getDestination.

bziemons committed:
Move TickExecutor init before server start

Fix #1430

bziemons committed:
Fix #1454

remove Railcraft support. getCraftMatrix needs a Container and links
it via its listener now. Cannot use that.

bziemons committed:
Refactor AsyncQuicksortModule

bziemons committed:
Stay synchronous in async quick sort, if possible

bziemons committed:
Stay synchronous in async extractor, if possible

bziemons committed:
Remove single entry abstractmodules package

and move LogisticsModule to logisticspipes.modules.

bziemons committed:
Adding an async threshold config option

Finding easy solutions for complex problems. The default value is a guess

bziemons committed:
Make roomForItem method with ItemStack

Deprecates roomForItem with ItemIdentifier.

bziemons committed:
Use new roomForItem method, remove depr. methods

Fix a lot of weird destination item count calculations.

bziemons committed:
Add async advanced extractor module

bziemons committed:
Delombok for Kotlin!

bziemons committed:
Add base async computer quicksort class

bziemons committed:
Propagate service reg. in extension modules

bziemons committed:
Remove code of old CC based modules

bziemons committed:
Replace modules with async alternatives

bziemons committed:
Fix slot to inventory comparison in extractor

bziemons committed:
Remove isServer within handleTileReachedServer

bziemons committed:
Clean up in handleTileReachedServer

Extract duplicate code. Remove useless code.

Bernasss12 committed:
Drawing System (WIP)
